### ISSUE TITLE
Bio.SubsMat.MatrixInfo "grant" matrix lacking K amino acid 

### ISSUE THREAD

**Issue by dshivak** [biopython/biopython#308]
The Grantham physiochemical matrix (Science 185, 862-864 (1974)) is missing the 'K' amino acid from the matrix in the module Bio.SubsMat.MatrixInfo.

A complete copy of the matrix is available at:
http://www.genome.jp/aaindex/

**Comment by peterjc** [comment]
Nice job @darcymason - I didn't have a chance to look at this last month as I was on holiday, but you seem to have solved this.

Going back to the original reference, Table 2 from:

https://doi.org/10.1126/science.185.4154.862
https://www.ncbi.nlm.nih.gov/pubmed/4843792

<img width="1198" alt="screenshot 2018-08-22 10 37 45" src="https://user-images.githubusercontent.com/63959/44455991-58caf000-a5f7-11e8-9dc5-4f7fa1120076.png">

This seems to match the table given in wikipedia:

https://en.wikipedia.org/wiki/Amino_acid_replacement#Grantham's_distance

However, with Biopython @idoerg followed the table here:

https://web.archive.org/web/19991111011852/http://www.embl-heidelberg.de:80/~vogt/matrices/grant.cmp

Which says:

```
R. Grantham, Amino Acid Difference Formula to Help Explain Protein Evolution,

Science, 185, 862-864




 Diagonal: Min: 215.0  Max: 215.0  Mean:215.00  Sigma: 0.00
 Rest:     Min:   0.0  Max: 210.0  Mean:115.03  Sigma:47.43
 Matrix:   Min:   0.0  Max: 215.0  Mean:124.55  Sigma:53.87

   S   R   U   P   E   L   V   Y   W   F   O   C   H   Q   N   I   A   G   M   T ..
 215 105  70 141 157 116  91 159  73  60  71 103 126 147 169  94 150 135  80  38   S
     215 113 112 144 103 119  90 118 118 138  35 186 172 129 189 119 161 124 114   R
         215 117 123 119 183  77 210 193 179  17 116 102  62 108  43  77 200 154   U
             215 177 188 147 173 120 101 105  46 138 139 124 112 107 122 128  68   P
                 215 157 146 156 126 112 123  66 168 173 150 137 130 150 134  87   E
                     215 151 155 121 102 103  20 129 124 104 109  89 108 131  67   L
                         215 106 186 165 160  23 131 119  82 118  63  94 194 127   V
                             215  80  62  68  56 117 128 135  88 121 117  88  31   Y
                                 215 194 182  17 121 106  66 113  47  81 205 154   W
                                     215 193  10 115  99  57 113  38  75 187 175   F
                                         215  21 132 116  72 130  55  93 179 178   O
                                             215  41  61  76  13  61  45  19   0   C
                                                 215 191 147 183 134 175 128 100   H
                                                     215 169 162 154 186 114  85   Q
                                                         215 121 192 173  73  41   N
                                                             215 114 159 120 105   I
                                                                 215 170  55  34   A
                                                                     215  89  63   G
                                                                         215 148   M
                                                                             215   T
```

There is also a summary listing of the tables here:


https://web.archive.org/web/19991014010917/http://www.embl-heidelberg.de:80/~vogt/matrices/mlist1.html

From inspection, the header, and the summary table, the min is 0 (for Threonine/Thr/T and Cysteine/Cys/C) and the max is 215 (self comparisons).

Notice here Leucine/Leu/L with Isoleucine/Ile/I is 109, while Cysteine/Cys/C with Tryptophan/Trp/W is 17, which are neither the highest or the lowest scores.

The values match those currently in Biopython, ``('I', 'L'): 109, `` and  ``('C', 'W'): 17,`` from here:

https://github.com/biopython/biopython/blob/biopython-172/Bio/SubsMat/MatrixInfo.py#L1732

Biopython has simply reflected that - right back to the first commit of the matrix by @idoerg:

https://github.com/biopython/biopython/commit/f223c67bc8a02289d1dcaa1d2b3d7edd48a7742b#diff-366000c21e8d3a02034af9a476387c1bR1555

Meanwhile the original and wikipedia have  Leucine/Leu/L and Isoleucine/Ile/I is 5, while Cysteine/Cys/C and Tryptophan/Trp/W is 215.

I looked at those numbers because wikipedia says: *According to Grantham's distance, most similar amino acids are leucine and isoleucine and the most distant are cysteine and tryptophan.* which does not match this at all.

So you've found a second problem - http://www.embl-heidelberg.de:80/~vogt/matrices/grant.cmp is missing the Lysine/Lys/K values, _and_ is inconsistent with the original table.

The worrying possibility is Vogt made a conversion error on top of dropping Lysine/Lys/K... and your explanation seem to be correct.

This is a screenshot from Excel, starting with the original Grant values (copied from Wikipedia, distances with 0 for self-matches), and then below that showing them rescaled to similarity scores of 215 for self match (using 215 minus the distance score).

<img width="1330" alt="screenshot 2018-08-22 11 30 30" src="https://user-images.githubusercontent.com/63959/44458770-2624f580-a5ff-11e8-8e15-ee8dd64db127.png">

As you noticed, the numbers now match Vogt's Grant table - but with the column labels in a different order. He would have added the diagonal for self matches, but appears to have mixed up the labels as part of that reformatting and omitted Lysine/Lys/K - and it seems to have taken 20 years for anyone to notice.

The good news is we presumably can now recalculate the intended values for Vogt's Grant table (scores up to 215), and update Biopython.

I don't see the tables on https://www.genome.jp/aaindex/ suggested by @dshivak which would be useful as a second opinion.

Unfortunately, from the old webpages I don't see any reference to a publication from Vogt which would be worth checking (and then potentially asking the editors to add a correction to).

I think our next step is to try to contact Vogt, although that might be difficult given they are likely no longer at EMBL Heidelberg - the last archive of the webpages was from 2000.

@idoerg - could you look at this please?

**Comment by darcymason** [comment]
@peterjc, I'm glad you were able to confirm this, but I should have saved you some time and attached some of the tables as you did (I had a spreadsheet very similar to what you pasted).

> I don't see the tables on https://www.genome.jp/aaindex/ suggested by @dshivak which would be useful as a second opinion.

The last part of my post above has a link to that table, and it is in the code I attached as well.  The order was completely different, so the code turns them into SeqMat tables and prints with the alphabets to make them match.

**Comment by peterjc** [comment]
Oh yeah, I missed the https://www.genome.jp/dbget-bin/www_bget?aaindex:GRAR740104 link and confess I didn't look at your code - in part I wanted to see if I came to the same conclusion as you without following the exact same route.

Please go ahead with the pull request correcting the dictionary in Biopython.

Looking at https://web.archive.org/web/19990220002158/http://www.embl-heidelberg.de:80/~vogt/rightframe.html gives us a full name and some (old) contact information for Gerhard Vogt if we do want to try to let him know as a curtesy.

**Comment by darcymason** [comment]
> Please go ahead with the pull request correcting the dictionary in Biopython.

Just to be clear -- should it be using the IUPAC characters?  And subtracted from 215?

**Comment by peterjc** [comment]
I don't understand your phrasing "using the IUPAC characters" as all of this is using IUPAC protein codes, just in different orders (and sometimes missing K).

You've convinced me that Gerhard Vogt made an error in computing his table which we copied, so yes - repeat the calculation of 215 less values in the paper/wikipedia, but this time with the correct amino acid labels.

**Comment by darcymason** [comment]
> I don't understand your phrasing "using the IUPAC characters" as all of this is using IUPAC protein codes, just in different orders (and sometimes missing K).

I should have been clearer about that.  I forgot to mention that there are actually two differences between the characters used:

```python
>>> vogt =  set("SRUPELVYWFOCHQNIAGMT")
>>> iupac = set("SRLPTAVGIFYCHQNKDEMW")
>>> differences = iupac ^ vogt
>>> differences & vogt
{'U', 'O'}
>>> differences & iupac
{'D', 'K'}
```
So vogt is missing both K and D, and has instead a U and O.  And many of the other characters in the  table row/col headers line up, but not all, so it does not seem like a simple conversion error that Vogt made -- I just wanted to make sure that there isn't something else going on, since I have no access to the original paper.

And then asking about the 215 subtraction: I wondered if it still made sense to still do this. Instead the unchanged aaindex/Wikipedia Grantham distance table could be used, and there would be no tie any more to the Vogt work.  But I'm not clear on how these numbers are used and whether users want the inverted numbers.

I have prepared the pull request (with the subtract from 215) and will submit that shortly,  Just ignore my comments above if they are not relevant.

**Comment by peterjc** [comment]
Looking at your pull request, from the test output I immediately spotted that we were missing 'K' and 'D', you're ahead of me with the set example showing it had instead 'U' and 'O'. We may never understand how the mix up happened, but your pull request does seem to be the original intent.

**Comment by peterjc** [comment]
Closed via #1781, thank you!
